New back-end algorithm RHO for findHomography()

obilaniu opened the pull request:
Implements the RHO algorithm for rapid homography estimation as presented in

Paper: Bilaniuk, Olexa, Hamid Bazargani, and Robert Laganiere. "Fast Target Recognition on Mobile Devices: Revisiting Gaussian Elimination for the Estimation of Planar Homographies." In Computer Vision and Pattern Recognition Workshops (CVPRW), 2014 IEEE Conference on, pp. 119-125. IEEE, 2014.

vpisarev commented on modules/calib3d/src/rhorefc.cpp:
```diff
@@ -0,0 +1,2165 @@
+ * IEEE, 2014.
+ */
+
+/* Includes */
```
precomp.hpp should be the first included header

vpisarev commented on modules/calib3d/src/rhorefc.cpp:
```diff
@@ -0,0 +1,2165 @@
+
+
+/* For the sake of cv:: namespace ONLY: */
+#ifdef __cplusplus
```
#ifndef __cplusplus should be omitted, all the OpenCV code is built with C++ compiler

vpisarev commented on modules/calib3d/src/rhorefc.cpp:
```diff
@@ -0,0 +1,2165 @@
+
+
+/* Defines */
+#define MEM_ALIGN               32
```
using defines is strongly discouraged; please, use const's.

vpisarev commented on modules/calib3d/src/rhorefc.cpp:
```diff
@@ -0,0 +1,2165 @@
+int  rhoRefCInit(RHO_HEST_REFC* p){
+    memset(p, 0, sizeof(*p));
+
+    p->ctrl.smpl   = (unsigned*)almalloc(m*sizeof(*p->ctrl.smpl));
```
please, use std::vector instead of plain buffers.

vpisarev commented on modules/calib3d/src/rhorefc.cpp:
```diff
@@ -0,0 +1,2165 @@
+ * Writes: eval.A, eval.lambdaAccept, eval.lambdaReject
+ */
+
+static inline void   sacDesignSPRTTest(RHO_HEST_REFC* p){
```
all such functions are better be shaped as RHP_HEST_REFC methods

vpisarev commented:
@obilaniu, thanks a lot for the contribution! I would like to ask you, though, to refactor/restyle the code a little bit to bring it to our current C++ style (see above comments). It's not our caprice, we will have to maintain this code for years, so it's much easier when it has the same style.

obilaniu commented:
@vpisarev Thank you for the feeback; There remains much work to be done, evidently. The code was written in C originally, whence comes its heavy use of idiomatically C constructs.

We are currently dealing with the accuracy-test failures on a separate branch so as to avoid abusing your buildbots. We'll also address your latest concerns, then make a new merge into our master branch (the one feeding the pull-request).

This is intented to be a much faster, though perhaps less accurate, homography estimator suitable for use on mobile devices or anywhere where speed is of serious concern.

vpisarev commented:
ok, thank you! So, does it mean you submit a new PR, or you are going to update this PR later?

obilaniu commented:
@vpisarev I will today push a commit that converts to a much more C++-like form the code; It addresses most of the comments you had above, though it will still fail the accuracy tests. We'll keep updating this PR for now.

obilaniu commented:
@vpisarev The code has been updated to address most of your C++ feedback. Is it a step in the right direction?

It still fails the homography accuracy test; We are at work on that problem.

obilaniu commented:
@vpisarev, my colleague @hamidb has corrected the accuracy problems, and as a result all tests have passed and all buildbots are green. May you review the code once more?

vpisarev commented:
@obilaniu, sorry for delay, I will review the code shortly

obilaniu commented:
@vpisarev Thank you for the second glance.

At present what you see is only the C++ reference implementation; We are considering (re-)introducing vectorized (SSE/AVX/AVX2/FMA, possibly NEON)-optimized versions. In the past I would simply copy the source code file whole, make the data structure alterations that are required and add the vectorized code; however, this would entail much code duplication. Is there a recommended way to maintain multiple optimized implementations of an algorithm alongside the reference one?

I currently favour a very coarse-grain #if-#endif approach and subclasses of a base class, for several reasons:
- Extensive source code changes are required to run the 4 or 8 PROSAC iterations in lockstep, as I intend to do. The changes are not just vectorizations of hotspots; They affect control flow and require changes to the context struct (see below).
- The homography estimation context structure must be changed in a number of places to support vectorized execution. In particular, a new LUT is required to implement the vectorized SPRT test; the PRNG, points buffers and homography buffers must be duplicated; and more.
- The reference C implementation is currently very readable and understandable. Peppering it with fine-grained #if-#endifs detracts from that quality.

How does OpenCV resolve this type of problem? We may put this refactoring into the future, but it should eventually be done. We are willing to proceed with just the plain C++ implementation for 3.0.0's initial releases, but feel that a full complement of optimized routines best demonstrates the sheer speed of our algorithm.

vpisarev commented on modules/calib3d/src/rhorefc.cpp:
```diff
@@ -0,0 +1,2426 @@
+#include "rhorefc.h"
+
+/* Defines */
+const int    MEM_ALIGN              = 32;
```
the constants should be put inside cv namespace

vpisarev commented on modules/calib3d/src/rhorefc.cpp:
```diff
@@ -0,0 +1,2426 @@
+ * Prototypes for purely-computational code.
+ */
+
+static inline void*  almalloc(size_t nBytes);
```
manual memory allocation/deallocation should be completely eliminated

obilaniu replied:
@vpisarev What do you suggest for aligned memory allocation? I wrote that allocator for 32-byte aligned memory because I foresee adding vectorized code and would like to use aligned loads/stores.

vpisarev commented on modules/calib3d/src/rhorefc.cpp:
```diff
@@ -0,0 +1,2426 @@
+ */
+
+RHO_HEST_REFC* rhoRefCInit(void){
+    RHO_HEST_REFC* p = new RHO_HEST_REFC;
```
manual memory management should be completely eliminated. use Ptr&lt;&gt;

vpisarev commented on modules/calib3d/src/rhorefc.cpp:
```diff
@@ -0,0 +1,2426 @@
+    lm.Jte      = NULL;
+
+#ifdef _WIN32
+    fastSeed(rand());
```
please, do not use rand(), random() etc. anywhere except samples. This will give different results on different architectures. Results must be deterministic.

obilaniu replied:
@ I'm seeding a (deterministic) PRNG called xorshift128+, which although it has lower period than many PRNGs, is extremely fast, and multiple instances can be run in parallel. We have thrown out calls to rand()/random() everywhere else in the code because of its large overhead, but because findHomography() provides no seed through its arguments we must source entropy somewhere.

Do you suggest seeding xorshift128+ as fastSeed((int)cv::theRNG()) or something similar?

vpisarev commented on modules/calib3d/src/rhorefc.cpp:
```diff
@@ -0,0 +1,2426 @@
+ */
+
+inline int    RHO_HEST_REFC::initialize(void){
+    ctrl.smpl   = (unsigned*)almalloc(SMPL_SIZE*sizeof(*ctrl.smpl));
```
please, switch completely to std::vector, cv::Mat, cv::Matx for arrays. cv::Ptr&lt;&gt; for pointers to classes/structures

vpisarev commented on modules/calib3d/src/rhorefc.cpp:
```diff
@@ -0,0 +1,2426 @@
+
+inline void   RHO_HEST_REFC::finalize(void){
+    if(init){
+        alfree(ctrl.smpl);
```
all the uses of almalloc/alfree should be eliminated

vpisarev commented on modules/calib3d/src/rhorefc.cpp:
```diff
@@ -0,0 +1,2426 @@
+        alfree(best.H);
+        alfree(curr.pkdPts);
+
+        memset(this, 0, sizeof(*this));
```
this is bad practice, this may include pointers to virtual tables etc. do not use memset() with this

vpisarev commented on modules/calib3d/src/rhorefc.cpp:
```diff
@@ -0,0 +1,2426 @@
+            int inList;
+
+            do{
+                currentSample[i] = (unsigned)(dataSetSize*fastRandom());
```
fastRandom should be replaced with OpenCV's RNG.

vpisarev commented on modules/calib3d/src/rhorefc.cpp:
```diff
@@ -0,0 +1,2426 @@
+    float f=H[0]+H[1]+H[2]+H[3]+H[4]+H[5]+H[6]+H[7];
+
+    /* degenerate = isnan(f); */
+    degenerate = f!=f;/* Only NaN is not equal to itself. */
```
this is very unsafe. Please, use cvIsNaN() instead

obilaniu replied:
@vpisarev I will make the change that you request here, but I object to it being called "very unsafe"; It is in fact the safest alternative short of using the C/C++ Standard's `isnan()`. The IEEE-754 standard requires this behaviour, so unless the compiler is noncompliant in that respect, it's 100% safe, besides being more portable than the implementation of `cvIsNaN()` which I reviewed. A member of the IEEE-754 committee had [this to say](http://stackoverflow.com/a/1573715/2809095) on the topic on StackOverflow.

vpisarev commented on modules/calib3d/src/rhorefc.cpp:
```diff
@@ -0,0 +1,2426 @@
+     */
+
+    scalar1=minor[0][0];
+    major[0][0]/=scalar1;
```
I'm not sure if compiler can optimize out the repetitive divisions by the same number. It's better to precompute 1.f/scalar1 and then multiply the results by it. Division is an order of magnitude slower operation than multiplication.

vpisarev commented on modules/calib3d/src/rhorefc.h:
```diff
@@ -0,0 +1,283 @@
+/* Defines */
+
+/* C++ does not have the restrict keyword. */
+#ifdef restrict
```
I suggest to remove "restrict" completely

vpisarev commented on modules/calib3d/src/rhorefc.h:
```diff
@@ -0,0 +1,283 @@
+ */
+
+/* Include Guards */
+#ifndef __RHOREFC_H__
```
the guarding macro should have OPENCV prefix to minimize probability of compile errors.

vpisarev commented on modules/calib3d/test/test_homography.cpp:
```diff
@@ -144,7 +144,7 @@ void CV_HomographyTest::print_information_1(int j, int N, int _method, const Mat
     cout << "Type of srcPoints: "; if ((j>-1) && (j<2)) cout << "Mat of CV_32FC2"; else  cout << "vector <Point2f>";
```
I have not tried this code and these tests myself, but I've got an impression that it prints a lot of info even if the tests succeed. Can we calm down the tests when it's all ok?

obilaniu replied:
@vpisarev Pardon? When the tests succeed they print almost nothing at all.

{code}
[----------] 1 test from Calib3d_StereoSGBM (1112 ms total)

[----------] 1 test from Calib3d_Homography
[ RUN      ] Calib3d_Homography.accuracy
[       OK ] Calib3d_Homography.accuracy (4533 ms)
[----------] 1 test from Calib3d_Homography (4533 ms total)

[----------] 6 tests from OCL_StereoMatcher/StereoBMFixture
{code}

vpisarev replied:
oh, ok, false alarm then, sorry :]

vpisarev commented:
@obilaniu, I've put a bunch of comments regarding the code. For now the further optimization is out of scope, we shall concentrate on the style first. Once code gets into OpenCV, it stays there for years, so while I'd prefer the code to match opencv style as close as possible and thus make it easier to maintain. adding optimization can be the next step

obilaniu commented:
@vpisarev Thank you for the review. I've already dropped a few replies below your in-line comments by way of explanation. In particular, I'd like to assure you that the algorithm, while it was previously nondeterministic, _is_ now as deterministic as the seed to the API fastSeed(). This was in fact inspired by an earlier pull request that did something similar for solvePnPRansac().

I'll take care of fixing memory management preferably by tomorrow.

hamidb commented:
@vpisarev Thank you for comments and review. I am also a bit lost about deterministic random function. 
@obilaniu I can address those comments. However, for the sake of avoiding conflicts I linger commiting new changes. Let me know if you need my help.

obilaniu commented:
@vpisarev  @hamidb I've committed most straightforward changes. What remains is converting to automatic memory management. This will take me more time.

obilaniu commented:
@vpisarev @hamidb I am done applying your feedpack from the second review, as I have now expunged all manual memory management and in its stead substituted cv::Mat for internal buffers and cv::Ptr for the external interface's opaque pointer to the H-estimation context.

There is what I believe to be a spurious failure, only on the Mac build, of the stitching performance test. Since our code has nothing to do with it, I believe this test failure can be ignored, though I'll attempt to make a worthwhile commit to trigger a rebuild.

I hope you will be much more satisfied when you review the code the next time. I've added a base class and made the reference C implementation as a derived class; Once code style is ironed out it will be easy to add vectorized versions of our code.

vpisarev commented on modules/calib3d/src/rho.cpp:
```diff
@@ -0,0 +1,2643 @@
+    /* Initialize it. */
+    if(p){
+        if(!p->initialize()){
+            p = Ptr<RHO_HEST>((RHO_HEST*)NULL);
```
use p.release() here

vpisarev commented on modules/calib3d/src/rho.cpp:
```diff
@@ -0,0 +1,2643 @@
+    nr.beta     = 0.0;
+
+
+    fastSeed((unsigned)cv::theRNG());
```
fastSeed() should be given a constant, not a random number, otherwise results will be different each time you initialize the class.

obilaniu commented:
@vpisarev I've made the changes you requested almost immediately, but forgot to ping you to report having done so.  Now that I'm correctly releasing the smart pointer and am seeding with a constant value, are there any further changes to be made to the PR?

vpisarev commented:
terribly sorry for delay. Looks great! Let's put it in! :+1:
